Sources/System/IORing.swift

```diff
@@ -0,0 +1,657 @@
+
+    public struct RegisteredResources<T>: RandomAccessCollection {
+        let resources: [T]
+
+        public var startIndex: Int { 0 }
+        public var endIndex: Int { resources.endIndex }
+        init(resources: [T]) {
+            self.resources = resources
+        }
+        public subscript(position: Int) -> IOResource<T> {
+            IOResource(resource: resources[position], index: position)
+        }
+        public subscript(position: UInt16) -> IOResource<T> {
+            IOResource(resource: resources[Int(position)], index: Int(position))
+        }
+    }
+
+    public var registeredBuffers: RegisteredResources<IORingBuffer.Resource> {
+        RegisteredResources(resources: _registeredBuffers ?? [])
+    }
+
+    public func unregisterBuffers() {
+        fatalError("failed to unregister buffers: TODO")
+    }
+
+    public func submitPreparedRequests() throws {
+        try _submitRequests(ring: submissionRing, ringDescriptor: ringDescriptor)
+    }
+
+    public func submitPreparedRequestsAndConsumeCompletions(
+        minimumCount: UInt32 = 1,
+        timeout: Duration? = nil,
+        consumer: (consuming IOCompletion?, IORingError?, Bool) throws -> Void
+    ) throws {
+        //TODO: optimize this to one uring_enter
+        try submitPreparedRequests()
+        try blockingConsumeCompletions(
+            minimumCount: minimumCount, 
+            timeout: timeout, 
+            consumer: consumer
+        )
+    }
+
+    public mutating func prepare(request: __owned IORequest) -> Bool {
+        var raw: RawIORequest? = request.makeRawRequest()
+        return _writeRequest(
+            raw.take()!, ring: &submissionRing, submissionQueueEntries: submissionQueueEntries)
+    }
+
+    mutating func prepare(linkedRequests: some BidirectionalCollection<IORequest>) {
+        guard linkedRequests.count > 0 else {
+            return
+        }
+        let last = linkedRequests.last!
+        for req in linkedRequests.dropLast() {
+            var raw = req.makeRawRequest()
+            raw.linkToNextRequest()
+            _writeRequest(
+                raw, ring: &submissionRing, submissionQueueEntries: submissionQueueEntries)
+        }
+        _writeRequest(
+            last.makeRawRequest(), ring: &submissionRing,
+            submissionQueueEntries: submissionQueueEntries)
+    }
+
+    //@inlinable //TODO: make sure the array allocation gets optimized out...
+    public mutating func prepare(linkedRequests: IORequest...) {
+        prepare(linkedRequests: linkedRequests)
+    }
+
+    public mutating func submit(linkedRequests: IORequest...) throws {
+        prepare(linkedRequests: linkedRequests)
+        try submitPreparedRequests()
+    }
+
+    deinit {
+        munmap(ringPtr, ringSize)
+        munmap(
+            UnsafeMutableRawPointer(submissionQueueEntries.baseAddress!),
+            submissionQueueEntries.count * MemoryLayout<io_uring_sqe>.size
+        )
+        close(ringDescriptor)
+    }
```

Comment:
I'm not opposed to the `munmap` calls being here, but I don't think that `close` should be. It's generally not a good idea to clean up resources in deinit, even in the `~Copyable` case. Overwhelmingly the ring will be stored in a reference type of some kind, so the deinit-based cleanup just gets moved out one layer.

I think I'd pretty strongly prefer we put a `close` function here. I'm open to it being `consuming`, though that would also probably be inconvenient too. But generally we shouldn't be getting surprise closes.

This goes double given @CodaFi's note that we want to prevent classes of errors where we shut things down before the resources loaned to the kernel are actually cleaned up. Having an explicit close function gives us a hook on which we can hang that cleanup.

Reply:
What are the downsides to doing cleanup in deinit other than not being able to surface any errors? Are they worse than leaking resources?